crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -185,56 +238,1250 @@ impl<S: States> NextState<S> {
+                    .after(ApplyStateTransition::<S::RawState>::apply()),
+            );
+    }
+
+    fn register_state_exist_systems_in_schedule<T: SubStates<SourceStates = Self>>(
+        schedule: &mut Schedule,
+    ) {
+        let system = |mut parent_changed: EventReader<StateTransitionEvent<S::RawState>>,
+                      event: EventWriter<StateTransitionEvent<T>>,
+                      commands: Commands,
+                      current_state: Option<ResMut<State<T>>>,
+                      state_set: Option<Res<State<S::RawState>>>| {
+            if parent_changed.is_empty() {
+                return;
+            }
+            parent_changed.clear();
+
+            let new_state =
+                if let Some(state_set) = S::convert_to_usable_state(state_set.as_deref()) {
+                    T::exists(state_set)
+                } else {
+                    None
+                };
+
+            match new_state {
+                Some(value) => {
+                    if current_state.is_none() {
+                        internal_apply_state_transition(
+                            event,
+                            commands,
+                            current_state,
+                            Some(value),
+                        );
+                    }
+                }
+                None => {
+                    internal_apply_state_transition(event, commands, current_state, None);
+                }
+            };
         };
+
+        schedule
+            .add_systems(system.in_set(ApplyStateTransition::<T>::apply()))
+            .add_systems(
+                apply_state_transition::<T>.in_set(StateTransitionSteps::ManualTransitions),
+            )
+            .add_systems(
+                should_run_transition::<T, OnEnter<T>>
+                    .pipe(run_enter::<T>)
+                    .in_set(StateTransitionSteps::EnterSchedules),
+            )
+            .add_systems(
+                should_run_transition::<T, OnExit<T>>
+                    .pipe(run_exit::<T>)
+                    .in_set(StateTransitionSteps::ExitSchedules),
+            )
+            .add_systems(
+                should_run_transition::<T, OnTransition<T>>
+                    .pipe(run_transition::<T>)
+                    .in_set(StateTransitionSteps::TransitionSchedules),
+            )
+            .configure_sets(
+                ApplyStateTransition::<T>::apply()
+                    .in_set(StateTransitionSteps::DependentTransitions)
+                    .after(ApplyStateTransition::<S::RawState>::apply()),
+            );
+    }
+}
+/// Trait defining a state that is automatically derived from other [`States`].
+///
+/// A Sub State is a state that exists only when the source state meet certain conditions,
+/// but unlike [`ComputedStates`] - while they exist they can be manually modified.
+///
+/// The [`StateSet`] is passed into the `exist` method whenever one of them changes, and the
```

Comment:
This paragraph doesn't seem particularly useful to end users in its current location: it really surfaces some of the implementation details too soon.